### PR TITLE
[wasm] crypto deadlock fix

### PR DESCRIPTION
**Pull request by pavelsavara**
- more locking before manipulation with the channel state.
- more asserts about expected states
- removed extra console.debug logging
- renamed wait method to make it clearer what we are waiting for

Fixes crypto deadlock: [Log on main](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-bebe7b94e4e74d45b6/WasmTestOnBrowser-subtlecrypto-System.Security.Cryptography.Tests/1/console.55b88872.log?helixlogtype=result)

### CODE REVIEW

File: src/mono/wasm/runtime/workers/dotnet-crypto-worker.ts [hunk 1]
```diff
@@ -134,28 +140,24 @@ class ChannelWorker {
...
                 Atomics.store(this.comm, this.MSG_SIZE_IDX, 0);
                 this._change_state_locked(this.STATE_AWAIT);
-            } finally {
-                this._release_lock();
-            }
+            });
+            if (done) break;
 
-            this._wait(this.STATE_AWAIT);
+            this._wait_for_state_to_change__from(this.STATE_AWAIT);
```
Comment (radical) [inline src/mono/wasm/runtime/workers/dotnet-crypto-worker.ts:146]:
```suggestion
            this._wait_for_state_to_change_from(this.STATE_AWAIT);
```

File: src/mono/wasm/runtime/crypto-worker.ts [hunk 2]
```diff
@@ -299,33 +294,39 @@ class LibraryChannel {
...
                 response += this.read_from_msg(0, size_to_read);
 
                 // The response is complete.
+                state = Atomics.load(this.comm, this.STATE_IDX);
                 if (state === this.STATE_RESP) {
                     Atomics.store(this.comm, this.MSG_SIZE_IDX, 0);
-                    break;
+                    return true;
+                } else if (state !== this.STATE_RESP_P) {
+                    throw new Error(`Unexpectd state ${state}`);
```
Comment (radical) [inline src/mono/wasm/runtime/crypto-worker.ts:311]:
```suggestion
                    throw new Error(`Unexpected state ${state}`);
```

**Comment by pavelsavara** [comment]
CI test with multiple workitems on Helix is https://github.com/dotnet/runtime/pull/73629

**Review by maraf — APPROVED** [review]

File: src/mono/wasm/runtime/workers/dotnet-crypto-worker.ts [hunk 3]
```diff
@@ -134,28 +140,24 @@ class ChannelWorker {
                 // Reset the size and transition to await state.
                 Atomics.store(this.comm, this.MSG_SIZE_IDX, 0);
                 this._change_state_locked(this.STATE_AWAIT);
-            } finally {
-                this._release_lock();
-            }
+            });
+            if (done) break;
```
Comment (eerhardt) [inline src/mono/wasm/runtime/workers/dotnet-crypto-worker.ts:144]:
```suggestion
            if (done)
                break;
```

It's easier to read if the body of an if is on a separate line.

File: src/mono/wasm/runtime/crypto-worker.ts [hunk 4]
```diff
@@ -358,7 +359,16 @@ class LibraryChannel {
...
+    private using_lock(cb: Function) {
+        try {
+            this._acquire_lock();
+            return cb();
+        } finally {
+            this._release_lock();
+        }
+    }
+
+    private _acquire_lock() {
```
Comment (eerhardt) [inline src/mono/wasm/runtime/crypto-worker.ts:371]: The function is marked `private` - does it need to have an underscore prefix?

File: src/mono/wasm/runtime/crypto-worker.ts [hunk 5]
```diff
@@ -335,20 +337,19 @@ class LibraryChannel {
...
+                const state = Atomics.load(this.comm, this.STATE_IDX);
+                if (state == this.STATE_REQ_FAILED)
+                    throw new OperationFailedError(`Worker failed during ${msg} with state=${state}`);
 
-            if (is_ready(state))
-                return state;
+                if (is_ready(state))
+                    return true;
+            });
+            if (done) return;
```
Comment (eerhardt) [inline src/mono/wasm/runtime/crypto-worker.ts:352]: (nit) it is much more readable to put the body of the `if` on a new line.

File: src/mono/wasm/runtime/crypto-worker.ts [hunk 6]
```diff
@@ -299,33 +294,40 @@ class LibraryChannel {
     private read_response(): string {
         let response = "";
         for (; ;) {
-            const state = this.wait_for_state(state => state == this.STATE_RESP || state == this.STATE_RESP_P, "read_response");
-            this.acquire_lock();
-
-            try {
+            this.wait_for_state_change_to(state => state == this.STATE_RESP || state == this.STATE_RESP_P, "read_response");
+            let state;
```
Comment (eerhardt) [inline src/mono/wasm/runtime/crypto-worker.ts:298]: Does this variable need to be outside of the `using_lock` callback? It doesn't appear to be used outside of it.

File: src/mono/wasm/runtime/crypto-worker.ts [hunk 7]
```diff
@@ -358,7 +359,16 @@ class LibraryChannel {
         return String.fromCharCode.apply(null, slicedMessage);
     }
 
-    private acquire_lock() {
+    private using_lock(cb: Function) {
```
Comment (eerhardt) [inline src/mono/wasm/runtime/crypto-worker.ts:362]: I'm not a web dev, but is `cb` a "typical" name for callback? Can it be named `callback` or `action`? That seems to be more readable.

**Review by eerhardt — APPROVED** [review]
Overall this looks good to me. Thanks for fixing this!

I just had a few nit-ish comments.

**Pull request merged**